misk-action-scopes/src/main/kotlin/misk/scope/ActionScope.kt

```diff
@@ -47,10 +47,18 @@ class ActionScope @Inject internal constructor(
   fun <T> propagate(c: Callable<T>): Callable<T> {
     check(tls.get() != null) { "not running within an ActionScope" }
 
+    val currentThreadId = Thread.currentThread().id
     val currentScopedData = tls.get().toMap()
+
     return Callable {
-      enter(currentScopedData).use {
+      // If the original thread is the same as the thread that calls the Callable and we are already
+      // in scope, then there is no need to re-enter the scope.
+      if (currentThreadId == Thread.currentThread().id && inScope()) {
```

Comment:
technically thread IDs can be reused which would make this check incorrect.

> When a thread is terminated, this thread ID may be reused.

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Thread.html#getId()

Reply:
Perhaps instead create some UUID and store in a threadlocal and compare against _that_...or perhaps `inScope()` is a sufficient check (not sure, but worth asking)

Reply:
You raised a good point. I replaced thread id comparison with a scoped seed data comparison. If the seed data is equal across executions, we don't attempt to re-enter the scope. If the seed data differs, it assumes that you're attempting to enter a new action scope, so we'll let it fail (by calling `enter` in an illegal state).

Reply:
Never mind. I got feedback your idea of UUID is probably the best approach:
https://cash.slack.com/archives/C01M5M297T7/p1661312203488089?thread_ts=1661304731.766609&cid=C01M5M297T7